price-feeder/oracle/provider/osmosisv2.go

```diff
@@ -0,0 +1,415 @@
+			return nil, err
+		}
+		candlePrices[cp.String()] = prices
+	}
+
+	return candlePrices, nil
+}
+
+func (p *OsmosisV2Provider) getTickerPrice(key string) (types.TickerPrice, error) {
+	p.mtx.RLock()
+	defer p.mtx.RUnlock()
+
+	ticker, ok := p.tickers[key]
+	if !ok {
+		return types.TickerPrice{}, fmt.Errorf(
+			types.ErrTickerNotFound.Error(),
+			ProviderOsmosisV2,
+			key,
+		)
+	}
+
+	return ticker, nil
+}
+
+func (p *OsmosisV2Provider) getCandlePrices(key string) ([]types.CandlePrice, error) {
+	p.mtx.RLock()
+	defer p.mtx.RUnlock()
+
+	candles, ok := p.candles[key]
+	if !ok {
+		return []types.CandlePrice{}, fmt.Errorf(
+			types.ErrCandleNotFound.Error(),
+			ProviderOsmosisV2,
+			key,
+		)
+	}
+
+	return candles, nil
+}
+
+// SubscribeCurrencyPairs performs a no-op since the osmosis-api does not
+// have specific currency channels.
+func (p *OsmosisV2Provider) SubscribeCurrencyPairs(pairs ...types.CurrencyPair) error {
+	return nil
+}
+
+func (p *OsmosisV2Provider) messageReceived(
+	messageType int,
+	bz []byte,
+	pairs ...types.CurrencyPair,
+) {
+	if messageType != websocket.TextMessage {
+		return
+	}
+
+	// check if message is an ack first
+	if string(bz) == "ack" {
+		return
+	}
+
+	var (
+		messageResp map[string]interface{}
+		messageErr  error
+		tickerResp  OsmosisV2Ticker
+		tickerErr   error
+		candleResp  OsmosisV2Candle
+		candleErr   error
+	)
+
+	messageErr = json.Unmarshal(bz, &messageResp)
+
+	// Check the response for currency pairs that the provider is subscribed
+	// to and determine whether it is a ticker or candle.
+	for _, pair := range pairs {
+		osmosisV2Pair := currencyPairToOsmosisV2Pair(pair)
+		if msg, ok := messageResp[osmosisV2Pair]; ok {
+			switch v := msg.(type) {
+			// ticker response
+			case map[string]interface{}:
+				tickerString, _ := json.Marshal(v)
+				tickerErr = json.Unmarshal(tickerString, &tickerResp)
+				if tickerErr != nil {
+					p.logger.Error().
+						Int("length", len(bz)).
+						AnErr("ticker", tickerErr).
+						Msg("Error on receive message")
+
+					return
+				}
+				p.setTickerPair(
+					osmosisV2Pair,
+					tickerResp,
+				)
+				telemetryWebsocketMessage(ProviderOsmosisV2, MessageTypeTicker)
+
+				return
+
+			// candle response
+			case []interface{}:
+				// use latest candlestick in list
+				candleString, _ := json.Marshal(v[len(v)-1].(map[string]interface{}))
+				candleErr = json.Unmarshal(candleString, &candleResp)
+				if candleErr != nil {
+					p.logger.Error().
+						Int("length", len(bz)).
+						AnErr("candle", candleErr).
+						Msg("Error on receive message")
+
+					return
+				}
+				p.setCandlePair(
+					osmosisV2Pair,
+					candleResp,
+				)
+				telemetryWebsocketMessage(ProviderOsmosisV2, MessageTypeCandle)
+
+				return
+			}
+		}
+	}
+
+	p.logger.Error().
+		Int("length", len(bz)).
+		AnErr("message", messageErr).
+		Msg("Error on receive message")
+}
+
+// ping to check websocket connection
+func (p *OsmosisV2Provider) ping() error {
```

Comment:
I don't think the osmosis-api responds to the standard ping message (maybe it should???). https://github.com/umee-network/osmosis-api/blob/main/server/client.go#L32. However, this might still work - because it will successfully send a message - the server will receive it and ignore it. It probably also fails after the connection has been closed.

Reply:
Ping is typically used as a way to say "please keep me alive!", here we can just keep it as a "are you there?" message. We should add a test to make sure the ping fails if there is no active connection.

One thing to note @zarazan is that the websocket API currently doesn't disconnect anything after x amount of time. Can you add a 24 hour disconnect, with a limit of 250 current clients? (we've currently got 111 validators in the inactive set)

Reply:
ping method is no longer there after using the WebsocketController